src/markov/markov_approx.jl

```diff
@@ -219,3 +223,567 @@ function estimate_mc_discrete{T}(X::Vector{T})
+- `Nm::Integer` : number of grid points
+- `M::Integer` : number of variables (M=1 corresponds to AR(1))
+- `nSigmas::Real` : number of standard error determining end points of grid
+- `method::Even` : method for grid making
+
+##### Return
+
+- `y1D` : (M x Nm) matrix of variable grid
+
+"""
+function construct_1D_grid(Sigma::ScalarOrArray, Nm::Integer,
+                           M::Integer, nSigmas::Real, method::Even)
+    if M == 1
+        minSigmas = sqrt(Sigma)
+    else
+        minSigmas = sqrt(minimum(eig(Sigma)[1]))
+    end
+    y1Drow = collect(linspace(-minSigmas*nSigmas, minSigmas*nSigmas, Nm))'
+    y1D = repmat(y1Drow, M, 1)
+    return y1D
+end
+
+"""
+
+Return combinations of each column of matrix `A`.
+It is simiplifying `allcomb2` by using `gridmake` from QuantEcon
+
+##### Arguments
+
+- `A::Matrix` : (N x M) Matrix
+
+##### Returns
+
+- (N^M x M) Matrix, combination of each row of `A`.
+
+###### Example
+```jldoctest
+julia>allcomb3([1 4 7;
+                2 5 8;
+                3 6 9]) # numerical input
+27×3 Array{Int64,2}:
+ 1  4  7
+ 1  4  8
+ 1  4  9
+ 1  5  7
+ 1  5  8
+ 1  5  9
+ 1  6  7
+ 1  6  8
+ 1  6  9
+ 2  4  7
+ ⋮
+ 2  6  9
+ 3  4  7
+ 3  4  8
+ 3  4  9
+ 3  5  7
+ 3  5  8
+ 3  5  9
+ 3  6  7
+ 3  6  8
+ 3  6  9
+```
+
+"""
+allcomb3(A::Matrix) =
+    flipdim(gridmake(flipdim([A[:, i] for i in 1:size(A, 2)], 1)...) , 2)
+
+"""
+Compute a discrete state approximation to a distribution with known moments,
+using the maximum entropy procedure proposed in Tanaka and Toda (2013)
+
+```julia
+p, lambdaBar, momentError = discrete_approximation(D, T, TBar, q, lambda0)
+```
+
+##### Arguments
+
+- `D::Vector` : (N x 1) vector of grid points. K is the dimension of the
+        domain. N is the number of points at which an approximation
+        is to be constructed.
+- `T::Function` : A function handle which should accept arguments of dimension
+        (N x 1) vector and return an (L x N) matrix of moments evaluated at
+        each grid point, where L is the number of moments to be
+        matched.
+- `TBar::Vector` : (L x 1) vector of moments of the underlying distribution
+          which should be matched
+
+##### Optional
+
+- `q::Matrix` : (1 X N) vector of prior weights for each point in D. The
+        default is for each point to have an equal weight.
+- `lambda0::Vector` : (L x 1) vector of initial guesses for the dual problem
+              variables. The default is a vector of zeros.
+
+##### Returns
+
+- `p` : (1 x N) vector of probabilties assigned to each grid point in `D`.
+- `lambdaBar` : (L x 1) vector of dual problem variables which solve the
+                maximum entropy problem
+- `momentError` : (L x 1) vector of errors in moments (defined by moments of
+                  discretization minus actual moments)
+
+"""
+function discrete_approximation(D::Vector, T::Function, TBar::Vector,
+                                  q::Matrix=ones(N)'/N, # Default prior weights
```

Comment:
My vote would be to have `q` input as an `AbstractVector` of length `N` instead of a `1 x N` matrix. 

I think it is easier (for me at least) to just pass a vector and then let you reshape it inside the function. 

Same holds for `lambda0`